src/runtime_src/hip/core/event.cpp

```diff
@@ -4,66 +4,180 @@
 #include "event.h"
 
 namespace xrt::core::hip {
+event::
+event(std::shared_ptr<stream>&& s)
+  : command(std::move(s))
+{
+  ctype = type::event;
+}
+
+void
+event::
+record(std::shared_ptr<stream> s)
+{
+  cstream = std::move(s);
+  auto ev = std::dynamic_pointer_cast<event>(command_cache.get(static_cast<command_handle>(this)));
+  throw_invalid_handle_if(!ev, "event passed is invalid");
+  if (is_recorded()) {
+    // already recorded
+    //cstream->erase_cmd(ev); It is commented will be uncomment after stream PR is pushed
+  }
+  // update recorded commands list
+  //cstream->enqueue_event(std::move(ev));It is commented will be uncomment after stream PR is pushed
+  set_state(state::recorded);
+}
 
 bool
-command::
-submit()
+event::
+is_recorded()
 {
-    return true; //temporary
+  return get_state() >= command::state::recorded;
 }
 
 bool
-command::
-wait()
+event::
+query()
 {
-    return true; //temporary
+  for (auto it = recorded_commands.begin(); it != recorded_commands.end(); it++){
+    state command_state = (*it)->get_state();
+    if (command_state != state::completed){
+      return false;
+    }
+  }
+  return true;
 }
 
 bool
 event::
-submit()
+synchronize()
 {
-    return true; //temporary
+  for (auto it = recorded_commands.begin(); it != recorded_commands.end(); it++){
+    state command_state = (*it)->get_state();
+    if (command_state < state::completed){
+      (*it)->wait();
+      (*it)->set_state(state::completed);
+    }
+  }
+  set_state(state::completed);
+  for (auto it = chain_of_commands.begin(); it != chain_of_commands.end(); it++){
+    (*it)->submit(true);
+  }
+  return true;
 }
 
 bool
 event::
 wait()
 {
-    return true; //temporary
+  ctime = std::chrono::system_clock::now();
+  return synchronize();
+}
+
+bool
+event::
+submit(bool)
+{
+  return true;
+}
+
+std::shared_ptr<stream>
+event::
+get_stream()
+{
+  return cstream;
+}
+
+void
+event::
+add_to_chain(std::shared_ptr<command> cmd)
+{
+  // lock and add
+}
+
+void
+event::
+add_dependency(std::shared_ptr<command> cmd)
+{
+  // lock and add
+}
+
+float
+event::
+elapsedtimecalc (std::shared_ptr<command> end)
+{
+  auto duration = end->get_time() - this->get_time();
+  auto millis = std::chrono::duration_cast<std::chrono::milliseconds>(duration).count();
+  return millis;
 }
 
 kernel_start::
-kernel_start(function &f, void* args)
-{}
+kernel_start(std::shared_ptr<stream>&& s, std::shared_ptr<function> &&f, void** args)
+  : command(std::move(s))
+{
+  ctype = type::kernel_start;
+  //xrt::kernel k = f->get_kernel();
+  xrt::kernel k; // just for compilation purpose we have to get it from function.
+  const auto& m_arginfo = std::move(xrt_core::kernel_int::get_args(k));
+  size_t idx = 0;
+  for (auto itr = m_arginfo.begin(); itr != m_arginfo.end(); ++itr, ++idx) {
+    xrt_core::kernel_int::set_arg_at_index(r, (*itr)->index, args[idx], (*itr)->size);
+  }
```

Comment:
Probably something like
```suggestion
  for (const auto& arginfo : m_arginfo)
    xrt_core::kernel_int::set_arg_at_index(r, arginfo.index, *args++, arginfo.size);
```
?